### DISCUSSION TITLE
Invalid bit depth when converting to sRGB from RGB16 with writeToBuffer

### DISCUSSION THREAD

**Discussion by L3tum** [libvips/libvips#2979] [Q&A] [answered]
## Bug report

**Describe the bug**
When converting a source image, which is RGB16 colourspace and uchar format, to sRGB colourspace via the PHP-method `writeToBuffer`, libvips will return an error saying `invalid bit depth`.

It can be fixed by explicitly calling `->colourspace(Interpration::SRGB)` on the image beforehand.

This did not happen prior to 8.13.0, as far as we're aware. We cannot reproduce it locally with the prebuilt Windows binaries, unfortunately.

**To Reproduce**
1. Use image [logo](https://user-images.githubusercontent.com/9307432/183414530-d09c5f26-c6ea-4901-bffd-812c1403ab70.png)
2. `->writeToBuffer(".png", ['profile' => 'srgb'])`
3. See error

**Expected behavior**
The image should be converted without any issues.

**Actual behavior**
Prints error `invalid bit depth`.

**Environment**
(please complete the following information)
 - OS: Debian Bullseye
 - Vips: 8.13.0 Built from Source

**Additional context**
- We're currently still using the Vips-Extension for PHP and not the FFI version

**Comment by jcupitt** [comment, marked as answer]
Hi @L3tum,

I think this is expected behaviour. You have a 16-bit image and you're trying to save it with an 8-bit profile. The depths don't match, so PNG save throws an error. If this did work previously, it was a bug.

I would write:

```php
$data = $image->colourspace(Interpration::SRGB)->writeToBuffer(".png");
```

The `colourspace` operation will convert to 8-bit, so you can attach the srgb profile if you like. However you probably don't need it -- it'll usually be assumed.

**Reply by L3tum** [reply]
Does that mean that 
````php
$image->colourspace(Interpretation::SRGB)->writeToBuffer(".png")
````
is different from 
````php
$image->writeToBuffer(".png", ['profile' => 'srgb'])
````
? It was my understanding that the two are identical in function, but maybe I misinterpreted the docs.

**Reply by jcupitt** [reply]
Yes, the first one will convert to 8-bit srgb, then save.

The second will save the image as-is (so a 16-bit image stays 16-bit), but try to attach the built-in libvips 8-bit srgb profile.

**Reply by L3tum** [reply]
Alright, thanks for the explanation!

**Reply by L3tum** [reply]
Hey @jcupitt, I do have one final question. I've run the same conversion via `vipsthumbnail` and it appears that it automatically converts the PNG into SRGB colourspace and uchar format, without me needing to add any CLI options.
I'm wondering now if the PHP method `Image::thumnail_buffer` will do the same, meaning that we don't need to manually call `->colourspace` ourself?

**Reply by jcupitt** [reply]
Yes, `thumbnail` uses srgb colourspace for the shrink, ie. 8 bit (or float if you set the `linear` option).

It's also much faster than `resize` (if that's what you're doing):

```
$ time vips thumbnail wtc.jpg x.jpg 1000
real	0m0.264s
user	0m0.309s
sys	0m0.030s

$ time vips resize wtc.jpg x.jpg 0.1
real	0m0.830s
user	0m0.803s
sys	0m0.177s
```

**Reply by L3tum** [reply]
Cool, thank you!
